politeiawww/paywall.go

```diff
@@ -1,47 +1,256 @@
 package main
 
 import (
-	www "github.com/decred/politeia/politeiawww/api/v1"
+	"fmt"
+	"sync"
+	"time"
+
+	v1 "github.com/decred/politeia/politeiawww/api/v1"
 	"github.com/decred/politeia/politeiawww/database"
 	"github.com/decred/politeia/util"
 )
 
-// ProcessVerifyUserPaymentTx verifies that the provided transaction
+type paywallInfo struct {
+	address     string
+	amount      uint64
+	txNotBefore int64
+}
+
+const (
+	// paywallExpiry is the amount of time the server will watch a paywall address
+	// for transactions. It gets reset when the user logs in or makes a call to
+	// RouteVerifyUserPayment.
+	paywallExpiry = time.Hour * 24
+
+	// paywallCheckGap is the amount of time the server sleeps after polling for
+	// a paywall address.
+	paywallCheckGap = time.Second * 1
+)
+
+var (
+	paywallUsers map[uint64]paywallInfo
+	mutex        sync.RWMutex
+)
+
+func paywallHasExpired(txNotBefore int64) bool {
+	expiryTime := time.Unix(txNotBefore, 0).Add(paywallExpiry)
+	return time.Now().After(expiryTime)
+}
+
+func addUser(user *database.User) {
+	paywallUsers[user.ID] = paywallInfo{
+		address:     user.NewUserPaywallAddress,
+		amount:      user.NewUserPaywallAmount,
+		txNotBefore: user.NewUserPaywallTxNotBefore,
+	}
+}
+
+func (b *backend) derivePaywallInfo(user *database.User) (string, uint64, int64, error) {
+	address, err := util.DerivePaywallAddress(b.params,
+		b.cfg.PaywallXpub, uint32(user.ID))
+	if err != nil {
+		err = fmt.Errorf("Unable to derive paywall address #%v "+
+			"for %v: %v", uint32(user.ID), user.Email, err)
+	}
+
+	return address, b.cfg.PaywallAmount, time.Now().Unix(), err
+}
+
+func (b *backend) checkForPayments() {
+	minConfirmations := b.cfg.MinConfirmationsRequired
+
+	// Check new user payments.
+	for {
+		mutex.RLock()
+		for k, v := range paywallUsers {
+			user, err := b.db.UserGetById(k)
+			if err != nil {
+				log.Errorf("cannot fetch user by id %v: %v\n", k, err)
+				continue
+			}
+
+			if b.HasUserPaid(user) {
+				// The user could have been marked as paid by RouteVerifyUserPayment,
+				// so just remove him from the in-memory pool.
+				mutex.Lock()
```

Comment:
Looks like you are trying to upgrade a read lock to a read/write lock.  Go doesn't do that.  You'll have to RUnlock first